src/Microsoft.OData.Core/UriParser/Parsers/SelectExpandOptionParser.cs

```diff
@@ -102,34 +102,26 @@ internal sealed class ExpandOptionParser
         internal int MaxSearchDepth { get; set; }
 
         /// <summary>
-        /// Building off of a PathSegmentToken, continue parsing any expand options (nested $filter, $expand, etc)
-        /// to build up an ExpandTermToken which fully represents the tree that makes up this expand term.
+        /// Building off a PathSegmentToken, continue parsing any select options (nested $filter, $expand, etc)
+        /// to build up an SelectTermToken which fully represents the tree that makes up this select term.
         /// </summary>
-        /// <param name="pathToken">The PathSegmentToken representing the parsed expand path whose options we are now parsing.</param>
-        /// <param name="optionsText">A string of the text between the parenthesis after an expand option.</param>
-        /// <returns>The list of expand term tokens based on the path token, and all available expand options.</returns>
-        internal List<ExpandTermToken> BuildExpandTermToken(PathSegmentToken pathToken, string optionsText)
+        /// <param name="pathToken">The PathSegmentToken representing the parsed select path whose options we are now parsing.</param>
+        /// <param name="optionsText">A string of the text between the parenthesis after a select option.</param>
+        /// <returns>The select term token based on the path token, and all available select options.</returns>
+        internal SelectTermToken BuildSelectTermToken(PathSegmentToken pathToken, string optionsText)

```

Comment:
>SelectTermToken [](start = 17, length = 15)

Why does BuildSelectTermToken return a SelectTermToken, while BuildExpandTermToken returns a List<ExpandTermToken>?

Reply:
Because they are related to "*".

For `$expand=*`, it will go through all navigation properties and for each navigation property, it will create a ExpandTermToken.  So, `$expand=*` will be parsed as `$expand=NP1,NP2,NP3,...`
That's why it's a list.

For `$select=*`, it will output a `WildcardSelectTermToken`, never expand to all properties.  That's why it's a single token.

#Resolved.

Reply:
I'm not sure the semantics of parsing $expand=* into $expand=NP1,NP2,NP3,... is correct, in that there may be dynamic navigation properties that we would exclude. But, that's a separate issue.  For $select, we need to distinguish between * and the list of all properties (the former includes dynamic properties, the later does not). So, the added behavior is the correct behavior.

---
In reply to: [297359685](https://github.com/OData/odata.net/pull/1488#discussion_r297359685) [](ancestors = 297359685)